src/contrib/mlir/backend/pass/affine_lowerer.cpp

```diff
@@ -1135,10 +1135,23 @@ REWRITER(NGReluOp) {
   NGRAPH_CHECK(lhs.getType().isa<MemRefType>());
   Type elemTy = lhs.getType().dyn_cast<MemRefType>().getElementType();
 
+  auto origOperand = op->getOperands()[0];
+  auto origIntType = origOperand.getType()
+                         .cast<NGTensorType>()
+                         .getElementType()
+                         .dyn_cast<IntegerType>();
+
   AffineLoopNestBuilder(ivs, lbs, ubs, steps)([&] {
     Value val = iLHS(ivs);
     Value zero = createZeroConstant(elemTy);
-    iRes(ivs) = std_select(val > zero, val, zero);
+    if (origIntType && origIntType.isUnsigned()) {
```

Comment:
What does `val > zero` turn into? It would be better if this remained as `val > zero` and all of this signed/unsigned distinction were moved there. Otherwise it needs to be repeated everywhere there is a signed/unsigned distinction, as you have had to do. It's too easy to miss a comparison, or for one to be added again in the future.

Reply:
See this code:

https://github.com/llvm/llvm-project/blob/e25f01be0c4c4f0355ae645599b3f54acb712ddd/mlir/lib/Dialect/Affine/EDSC/Builders.cpp#L268

The intrinsic for operator> creates a signed comparison.  And, the Affine dialect is "signless" meaning that all integers are neither signed nor unsigned.  The signedess, in this case, is carried in the predicate for comparison:  signed greater than (sgt) vs. unsigned greater than (ugt).  It is the responsibility of the dialect lowering to Affine to pick the correct predicate.  Which mean, I think, that the if/else in the nGraph Affine lowerer here are necessary.  The unsigned lowering is a bit uglier as there is no intrinsic for it.  We could add it.  But, the Affine operator> will have to pick one way or the other (signed or unsigned) as it has no way to distinguish between signed and unsigned once lowered to Affine.

Your point about "too easy to miss a comparison" is a good one... I missed ArgMin and ArgMax with this commit.  Will follow up with a separate commit for those two ops.

Reply:
I think we should submit an MLIR PR to fix those operators.

Reply:
The fix would involve a redesign of the Affine dialect to use signed integers.  Are we going to hold up this PR waiting on such a change?

Reply:
Note that we have a correctness issue in our MLIR implementation at the moment.  Unsigned integers with the most significant bit set are treated as signed (negative) and therefore we are generating that incorrect answer.